Результаты по контрольным вопросам в конце файла
---

# Считываем файл в DataFrame



метод .read_csv(path,delimiter) 

In [9]:
import pandas as pd

In [10]:
df = pd.read_csv(r'D:\Study\IT-ml&ai-python\URSU\cars_moldova.csv', delimiter = ',')

In [11]:
df

Make         Model  Year      Style  Distance  \
0          Toyota         Prius  2011  Hatchback  195000.0   
1         Renault  Grand Scenic  2014  Universal  135000.0   
2      Volkswagen          Golf  1998  Hatchback       1.0   
3         Renault        Laguna  2012  Universal  110000.0   
4            Opel         Astra  2006  Universal  200000.0   
...           ...           ...   ...        ...       ...   
41002       Dacia     Logan Mcv  2015  Universal   89000.0   
41003     Renault         Modus  2009  Hatchback     225.0   
41004    Mercedes       E Class  2016      Sedan   50000.0   
41005       Mazda             6  2006      Combi  370000.0   
41006     Renault  Grand Scenic  2006    Minivan  300000.0   

       Engine_capacity(cm3)     Fuel_type Transmission  Price(euro)  
0                    1800.0        Hybrid    Automatic       7750.0  
1                    1500.0        Diesel       Manual       8550.0  
2                    1400.0        Petrol       Manual       2200.0  
3                    1500.0        Diesel       Manual       6550.0  
4                    1600.0  Metan/Propan       Manual       4100.0  
...                     ...           ...          ...          ...  
41002                1500.0        Diesel       Manual       7000.0  
41003                1500.0        Diesel       Manual       4500.0  
41004                1950.0        Diesel    Automatic      29500.0  
41005                2000.0        Diesel       Manual       4000.0  
41006                1500.0        Diesel       Manual       4000.0  

[41007 rows x 9 columns]

# Информация о DataFrame


Метод .info() 

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41007 entries, 0 to 41006
Data columns (total 9 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Make                  41007 non-null  object 
 1   Model                 41007 non-null  object 
 2   Year                  41007 non-null  int64  
 3   Style                 41007 non-null  object 
 4   Distance              41007 non-null  float64
 5   Engine_capacity(cm3)  41007 non-null  float64
 6   Fuel_type             41007 non-null  object 
 7   Transmission          41007 non-null  object 
 8   Price(euro)           41007 non-null  float64
dtypes: float64(3), int64(1), object(5)
memory usage: 2.8+ MB


# Поиск Дубликатов
 


методы .duplicated() и .sum()

In [13]:
df.duplicated().sum()

3743

**Удаление дубликатов**

Методы .drop_duplicates() и .reset_index()

In [14]:
DF = df.drop_duplicates()

DF = DF.reset_index(drop=True)

DF

Make         Model  Year      Style  Distance  \
0          Toyota         Prius  2011  Hatchback  195000.0   
1         Renault  Grand Scenic  2014  Universal  135000.0   
2      Volkswagen          Golf  1998  Hatchback       1.0   
3         Renault        Laguna  2012  Universal  110000.0   
4            Opel         Astra  2006  Universal  200000.0   
...           ...           ...   ...        ...       ...   
37259  Land Rover    Freelander  2002  Crossover  225000.0   
37260       Dacia     Logan Mcv  2015  Universal   89000.0   
37261     Renault         Modus  2009  Hatchback     225.0   
37262       Mazda             6  2006      Combi  370000.0   
37263     Renault  Grand Scenic  2006    Minivan  300000.0   

       Engine_capacity(cm3)     Fuel_type Transmission  Price(euro)  
0                    1800.0        Hybrid    Automatic       7750.0  
1                    1500.0        Diesel       Manual       8550.0  
2                    1400.0        Petrol       Manual       2200.0  
3                    1500.0        Diesel       Manual       6550.0  
4                    1600.0  Metan/Propan       Manual       4100.0  
...                     ...           ...          ...          ...  
37259                1800.0  Metan/Propan       Manual       4400.0  
37260                1500.0        Diesel       Manual       7000.0  
37261                1500.0        Diesel       Manual       4500.0  
37262                2000.0        Diesel       Manual       4000.0  
37263                1500.0        Diesel       Manual       4000.0  

[37264 rows x 9 columns]

# Индексация

**Визуализация части DataFrame**

Метод .head(n) - первые n строк

In [15]:
DF.head(6)

Make         Model  Year      Style  Distance  Engine_capacity(cm3)  \
0      Toyota         Prius  2011  Hatchback  195000.0                1800.0   
1     Renault  Grand Scenic  2014  Universal  135000.0                1500.0   
2  Volkswagen          Golf  1998  Hatchback       1.0                1400.0   
3     Renault        Laguna  2012  Universal  110000.0                1500.0   
4        Opel         Astra  2006  Universal  200000.0                1600.0   
5    Mercedes          Vito  2000   Microvan  300000.0                2200.0   

      Fuel_type Transmission  Price(euro)  
0        Hybrid    Automatic       7750.0  
1        Diesel       Manual       8550.0  
2        Petrol       Manual       2200.0  
3        Diesel       Manual       6550.0  
4  Metan/Propan       Manual       4100.0  
5        Diesel       Manual       3490.0

Метод .tail(n) - последние n строк

In [16]:
DF.tail(9)

Make         Model  Year      Style  Distance  \
37255     Renault        Megane  2011      Sedan  145000.0   
37256      Toyota         Auris  2007  Hatchback  222111.0   
37257       Skoda        Superb  2016  Cabriolet   78900.0   
37258  Volkswagen        Passat  2016      Sedan   88000.0   
37259  Land Rover    Freelander  2002  Crossover  225000.0   
37260       Dacia     Logan Mcv  2015  Universal   89000.0   
37261     Renault         Modus  2009  Hatchback     225.0   
37262       Mazda             6  2006      Combi  370000.0   
37263     Renault  Grand Scenic  2006    Minivan  300000.0   

       Engine_capacity(cm3)     Fuel_type Transmission  Price(euro)  
37255                1500.0        Diesel    Automatic       6200.0  
37256                1400.0        Diesel    Automatic       4444.0  
37257                2000.0        Petrol    Automatic      21000.0  
37258                1800.0        Petrol    Automatic      11500.0  
37259                1800.0  Metan/Propan       Manual       4400.0  
37260                1500.0        Diesel       Manual       7000.0  
37261                1500.0        Diesel       Manual       4500.0  
37262                2000.0        Diesel       Manual       4000.0  
37263                1500.0        Diesel       Manual       4000.0

**Индексация**

аттрибут .loc[строки, столбцы] 

In [17]:
DF.loc[1437,'Transmission']

'Automatic'

In [18]:
DF.loc[:,'Transmission']

0        Automatic
1           Manual
2           Manual
3           Manual
4           Manual
           ...    
37259       Manual
37260       Manual
37261       Manual
37262       Manual
37263       Manual
Name: Transmission, Length: 37264, dtype: object

In [19]:
DF.loc[:,['Transmission']].head(5)

Transmission
0    Automatic
1       Manual
2       Manual
3       Manual
4       Manual

In [20]:
DF.loc[:,['Transmission','Year']].head(5)

Transmission  Year
0    Automatic  2011
1       Manual  2014
2       Manual  1998
3       Manual  2012
4       Manual  2006

In [21]:
DF.loc[:,'Make':'Style'].head(5)

Make         Model  Year      Style
0      Toyota         Prius  2011  Hatchback
1     Renault  Grand Scenic  2014  Universal
2  Volkswagen          Golf  1998  Hatchback
3     Renault        Laguna  2012  Universal
4        Opel         Astra  2006  Universal

In [22]:
DF.loc[69,:]

Make                        Honda
Model                        CR-V
Year                         2002
Style                   Universal
Distance                 275100.0
Engine_capacity(cm3)       2000.0
Fuel_type                  Petrol
Transmission               Manual
Price(euro)                5900.0
Name: 69, dtype: object

In [23]:
DF.loc[69:69,:]

Make Model  Year      Style  Distance  Engine_capacity(cm3) Fuel_type  \
69  Honda  CR-V  2002  Universal  275100.0                2000.0    Petrol   

   Transmission  Price(euro)  
69       Manual       5900.0

In [24]:
DF.loc[322:1437,:]

Make     Model  Year      Style  Distance  Engine_capacity(cm3)  \
322      Skoda   Octavia  2010  Universal       1.0                1400.0   
323   Mercedes   E Class  2016      Sedan  161024.0                2200.0   
324      Honda     Civic  2006  Hatchback  195000.0                1400.0   
325       Ford    Fusion  2015      Sedan  124000.0                2000.0   
326       Ford    Fusion  2003  Hatchback  330000.0                1400.0   
...        ...       ...   ...        ...       ...                   ...   
1433   Peugeot       208  2013  Hatchback  154000.0                1400.0   
1434  Infiniti      QX30  2017        SUV   91000.0                1500.0   
1435      Ford      Kuga  2016        SUV  102500.0                1500.0   
1436      Ford      Kuga  2018        SUV  122100.0                1500.0   
1437       BMW  3 Series  2016      Sedan   64000.0                2000.0   

           Fuel_type Transmission  Price(euro)  
322           Petrol       Manual       5350.0  
323           Diesel    Automatic      26950.0  
324           Petrol       Manual       5400.0  
325   Plug-in Hybrid    Automatic      11499.0  
326           Diesel    Automatic       4100.0  
...              ...          ...          ...  
1433          Diesel    Automatic       6999.0  
1434          Diesel    Automatic      15490.0  
1435          Diesel       Manual      14490.0  
1436          Diesel       Manual      14250.0  
1437          Petrol    Automatic      17500.0  

[1116 rows x 9 columns]

In [25]:
DF.loc[227:229,'Make':'Fuel_type']

Make    Model  Year      Style  Distance  Engine_capacity(cm3)  \
227     Opel  Calibra  1992      Coupe  400000.0                2000.0   
228  Renault   Kadjar  2016  Crossover  150425.0                1500.0   
229     Ford   Fusion  2018      Sedan  105000.0                2000.0   

        Fuel_type  
227  Metan/Propan  
228        Diesel  
229        Hybrid

**Логическое Индексирование**

In [26]:
DF[DF['Transmission']=='Manual']

Make         Model  Year      Style  Distance  \
1         Renault  Grand Scenic  2014  Universal  135000.0   
2      Volkswagen          Golf  1998  Hatchback       1.0   
3         Renault        Laguna  2012  Universal  110000.0   
4            Opel         Astra  2006  Universal  200000.0   
5        Mercedes          Vito  2000   Microvan  300000.0   
...           ...           ...   ...        ...       ...   
37259  Land Rover    Freelander  2002  Crossover  225000.0   
37260       Dacia     Logan Mcv  2015  Universal   89000.0   
37261     Renault         Modus  2009  Hatchback     225.0   
37262       Mazda             6  2006      Combi  370000.0   
37263     Renault  Grand Scenic  2006    Minivan  300000.0   

       Engine_capacity(cm3)     Fuel_type Transmission  Price(euro)  
1                    1500.0        Diesel       Manual       8550.0  
2                    1400.0        Petrol       Manual       2200.0  
3                    1500.0        Diesel       Manual       6550.0  
4                    1600.0  Metan/Propan       Manual       4100.0  
5                    2200.0        Diesel       Manual       3490.0  
...                     ...           ...          ...          ...  
37259                1800.0  Metan/Propan       Manual       4400.0  
37260                1500.0        Diesel       Manual       7000.0  
37261                1500.0        Diesel       Manual       4500.0  
37262                2000.0        Diesel       Manual       4000.0  
37263                1500.0        Diesel       Manual       4000.0  

[20723 rows x 9 columns]

In [27]:
DF[(DF['Transmission']=='Manual') & (DF['Year']>2000)]

Make         Model  Year      Style  Distance  \
1         Renault  Grand Scenic  2014  Universal  135000.0   
3         Renault        Laguna  2012  Universal  110000.0   
4            Opel         Astra  2006  Universal  200000.0   
6      Volkswagen        Passat  2010  Universal  290000.0   
9             BMW      5 Series  2004      Sedan  325000.0   
...           ...           ...   ...        ...       ...   
37259  Land Rover    Freelander  2002  Crossover  225000.0   
37260       Dacia     Logan Mcv  2015  Universal   89000.0   
37261     Renault         Modus  2009  Hatchback     225.0   
37262       Mazda             6  2006      Combi  370000.0   
37263     Renault  Grand Scenic  2006    Minivan  300000.0   

       Engine_capacity(cm3)     Fuel_type Transmission  Price(euro)  
1                    1500.0        Diesel       Manual       8550.0  
3                    1500.0        Diesel       Manual       6550.0  
4                    1600.0  Metan/Propan       Manual       4100.0  
6                    1390.0        Petrol       Manual       5600.0  
9                     300.0        Diesel       Manual       3400.0  
...                     ...           ...          ...          ...  
37259                1800.0  Metan/Propan       Manual       4400.0  
37260                1500.0        Diesel       Manual       7000.0  
37261                1500.0        Diesel       Manual       4500.0  
37262                2000.0        Diesel       Manual       4000.0  
37263                1500.0        Diesel       Manual       4000.0  

[15562 rows x 9 columns]

# Сортировка DataFrame



Метод .sort_values(by)

In [28]:
DF.sort_values(by = 'Price(euro)').tail(5)

Make               Model  Year      Style  Distance  \
13131      Toyota  Land Cruiser Prado  2012        SUV   10005.0   
14834  Volkswagen               Caddy  2015      Combi  158260.0   
29132    Mercedes             G Class  2020        SUV    3000.0   
17849         KIA            Sportage  2018  Crossover   46000.0   
33677  Brilliance                 BS3  2009      Sedan   57000.0   

       Engine_capacity(cm3)     Fuel_type Transmission  Price(euro)  
13131                3000.0        Diesel    Automatic     200000.0  
14834                1395.0  Metan/Propan       Manual     200000.0  
29132                3982.0        Petrol    Automatic     215000.0  
17849                1600.0        Petrol    Automatic     490000.0  
33677                1598.0        Petrol       Manual   10000000.0


Метод .sort_values(by, ascending= False)

In [36]:
DF.sort_values(by = 'Year', ascending= True).head(5)

Make     Model  Year      Style   Distance  Engine_capacity(cm3)  \
16138       Volvo       S80  1900      Coupe  1556666.0                1212.0   
22563        Opel     Astra  1900  Universal        0.0                1000.0   
22577         BMW  7 Series  1900      Sedan        0.0                1000.0   
16043  Mitsubishi    Galant  1900    Minivan  1000000.0                2000.0   
22596     Renault        19  1900  Hatchback        0.0                 100.0   

      Fuel_type Transmission  Price(euro)  
16138    Petrol    Automatic      10000.0  
22563    Petrol       Manual       1500.0  
22577    Petrol    Automatic       1500.0  
16043    Petrol       Manual       7300.0  
22596    Petrol       Manual       1000.0

# Сохранение DataFrame в файл


Метод .to_csv(path, index) 

In [32]:
DF.to_csv(r'D:\Study\IT-ml&ai-python\URSU\cars_moldova_no_dup.csv',index=False)

# Контрольные вопросы

Вопрос: 
<b> Назовите тип данных категориальных столбцов в Pandas</b>

Ответ: <b>object</b>


In [55]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 254 entries, 0 to 253
Data columns (total 9 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Make                  254 non-null    object 
 1   Model                 254 non-null    object 
 2   Year                  254 non-null    int64  
 3   Style                 254 non-null    object 
 4   Distance              254 non-null    float64
 5   Engine_capacity(cm3)  254 non-null    float64
 6   Fuel_type             254 non-null    object 
 7   Transmission          254 non-null    object 
 8   Price(euro)           254 non-null    float64
dtypes: float64(3), int64(1), object(5)
memory usage: 18.0+ KB


Вопрос: 
<b>Назовите целевой столбец для задачи регрессии</b>

Ответ: <b>Цена автомобиля</b>

Вопрос: 
<b>Назовите целевой столбец для задачи классификации</b>

Ответ: <b>Тип трансмиссии в задании, но можно использовать и стиль, марку, тип потребляемого топлива</b>


Вопрос: 
<b>Возьмите датафрейм после удаления дубликатов. Выберете из полного датафрейма строки с индекса 69 по 322. Отсортируйте полученный датафрейм по колонке  'Distance' по убыванию. Какое значение колонки  'Style' у полученного датафрейма во второй строке сверху</b>

Ответ: <b>Minivan</b> 

In [52]:
df = pd.read_csv(r'D:\Study\IT-ml&ai-python\URSU\cars_moldova_no_dup.csv')
df = df.loc[69:322,:]
df = df.sort_values(by='Distance', ascending=False)
df = df.reset_index(drop=True)
df.loc[1:1, 'Style']

1    Minivan
Name: Style, dtype: object

Вопрос:
<b> Укажите правильную размерность исходного набора данных</b>

Ответ: <b>41007 строк × 9 колонок</b>